src/backend/metal/src/command.rs

```diff
@@ -1211,21 +1227,73 @@ impl CommandBuffer {
 
     fn set_depth_stencil_desc(
         &mut self,
-        depth_stencil_desc: &metal::DepthStencilState,
+        depth_stencil_desc: &pso::DepthStencilDesc,
+        depth_stencil_raw: &metal::DepthStencilState,
     ) -> soft::RenderCommand {
         self.state.depth_stencil_desc = Some(depth_stencil_desc.clone());
-        soft::RenderCommand::SetDepthStencilDesc(depth_stencil_desc.clone())
+        self.state.depth_stencil_raw = Some(depth_stencil_raw.clone());
+        soft::RenderCommand::SetDepthStencilDesc(depth_stencil_raw.clone())
     }
 
     fn set_stencil_reference_values(
         &mut self,
         front: pso::StencilValue,
         back: pso::StencilValue,
     ) -> soft::RenderCommand {
-        self.state.stencil_front_reference = front;
-        self.state.stencil_back_reference = back;
+        self.state.stencil.front_reference = front;
+        self.state.stencil.back_reference = back;
         soft::RenderCommand::SetStencilReferenceValues(front, back)
     }
+
+    fn set_stencil_mask_values(
+        &mut self,
+        front_back_read_masks: Option<(pso::StencilValue, pso::StencilValue)>,
+        front_back_write_masks: Option<(pso::StencilValue, pso::StencilValue)>,
+        depth_stencil_desc: Option<pso::DepthStencilDesc>,
+    ) -> Option<soft::RenderCommand> {
+        if let Some((f, b)) = front_back_read_masks {
+            self.state.stencil.front_read_mask = f;
+            self.state.stencil.back_read_mask = b;
+        }
+
+        if let Some((f, b)) = front_back_write_masks {
+            self.state.stencil.front_write_mask = f;
+            self.state.stencil.back_write_mask = b;
+        }
+
+        depth_stencil_desc.map(|desc| {
+            let mut new_desc = desc.clone();
+
+            let (mut front_face, mut back_face) = match new_desc.stencil {
+                pso::StencilTest::On { ref front, ref back } => {
+                    (front.clone(), back.clone())
+                }
+                pso::StencilTest::Off => {
+                    (pso::StencilFace::default(), pso::StencilFace::default())
+                }
+            };
+
+            if let Some((f, b)) = front_back_read_masks {
+                front_face.mask_read = pso::State::Static(f);
+                back_face.mask_read = pso::State::Static(b);
+            }
+
+            if let Some((f, b)) = front_back_write_masks {
```

Comment:
so I'm concerned about the following case:
  1. user calls function (Some, None). We update `self.state.stencil` fields for A, then create a new ds descriptor and produce a `SetDepthStencilDesc`
  2. user calls function (None, Some). We update fields for B (all good so far), then create a new ds descriptor that only has the values of B, where we need it to have both changes from the first and second calls by this time